### ISSUE TITLE
Good training and validation accuracy, poor testing accuracy

### ISSUE THREAD

**Issue by prokope3** [apache/incubator-mxnet#5228]
Hello

During training neural network for classification (3 classes)  I encountered very strange behavior of training. validation and testing errors.
All data is stored in single .csv file and is splitted before every training randomly into 3 parts (training, valiadation - used for early stopping, testing) in ratio 70:20:10. 

During training training accuracy goes up to 99,99 % and validation accuracy to 86%. My problem begins, when trained model is used on testing data, because it has only 44 % accuracy.

When I tried train different model (xgboost) validation and testing errors were roughly the same (as the should be in my opinion).

Any ideas, what could cause this behavior?

EDIT: Dataset has 288+1 attributes and 1,6M rows.

Code:
``` R
rmsprob <- mx.opt.create(learning.rate = 0.001, wd = 0.00001, gamma1 = 0.95, 
                         gamma2 = 0.9, name="rmsprop", rescale.grad = 1/512, clip_gradient=80)

net <- mx.symbol.Variable("data")
net <- mx.symbol.FullyConnected(net, num_hidden=512)
net <- mx.symbol.BatchNorm(net,eps=0.001,momentum=0.99,fix.gamma=T)
net <- mx.symbol.Activation(net, act_type="relu")
net <- mx.symbol.FullyConnected(net, num_hidden=512)
net <- mx.symbol.BatchNorm(net,eps=0.001,momentum=0.99,fix.gamma=T)
net <- mx.symbol.Activation(net, act_type="relu")
net <- mx.symbol.FullyConnected(net, num_hidden=512)
net <- mx.symbol.BatchNorm(net,eps=0.001,momentum=0.99,fix.gamma=T)
net <- mx.symbol.Activation(net, act_type="relu")
net <- mx.symbol.FullyConnected(net, num_hidden=3)
net <- mx.symbol.SoftmaxOutput(net)

model <- mx.model.FeedForward.create(net, X=train.x, y=train.y, initializer = mx.init.Xavier(), eval.data = list(data=test.x,  label=test.y),
                                     num.round=250, array.batch.size=512, verbose = TRUE, epoch.end.callback = early.stop(10), 
                                     optimizer =  rmsprob, eval.metric=mx.metric.accuracy, shuffle=TRUE) 

```

\# Custom version of mx.callback.early.stop
``` R
early.stop <- function(bad.steps = 5) {

    function(iteration, nbatch, env, verbose = TRUE) {
        
        
        # not worse than previous X steps
        if (!is.null(bad.steps)) {
            
            # set / reset iteration variables
            # it may be not the best practice to use global variables,
            # but let's not touch "model.r" file
            if (iteration == 1){
                # reset iterator
                mx.best.iter <<- 1
                
                # reset best score
                mx.best.score <<- 0
                
            }
            
            # test early stop round
            if (!is.null(env$eval.metric)) {
                
                result <- env$metric$get(env$eval.metric)
                
                if (result$value <= mx.best.score ) {
                    
                    if (mx.best.iter >= bad.steps) {
                        if (verbose) {
                            cat(paste0("Best score=", mx.best.score, ", iteration [", iteration - bad.steps, "] \n"))
                        }
                        return(FALSE)
                    } else {
                        mx.best.iter <<- mx.best.iter + 1
                    }
                    
                } else {
                    mx.best.score <<- result$value
                    # Keep just the best model
                    mx.model.save(env$model,"models/nnet",1)
                    mx.best.iter <<- 1
                }
            }
        }
        
        return(TRUE)
    }
}
```

prediction: 
``` R
load(file="clean_data/test.csv")
data <- data.frame(testing)
test.x <-  data.matrix(data[, -1])
test.y <-  as.numeric(as.factor(data[,1]))-1
preds <-  predict(model, test.x, array.layout = "rowmajor",
                  array.batch.size = 512 )
pred.label <- max.col(t(preds))-1
confusionMatrix(as.factor(pred.label), as.factor(test.y))
```

Operating System: Windows 10 (x64)
Compiler: x64
Package used (Python/R/Scala/Julia): R
MXNet version: 0.9.4
R `sessionInfo()`:
R version 3.3.2 (2016-10-31)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

locale:
[1] LC_COLLATE=Czech_Czech Republic.1250  LC_CTYPE=Czech_Czech Republic.1250   
[3] LC_MONETARY=Czech_Czech Republic.1250 LC_NUMERIC=C                         
[5] LC_TIME=Czech_Czech Republic.1250    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] caret_6.0-73      lattice_0.20-34   mxnet_0.9.4       dplyr_0.5.0      
[5] data.table_1.10.4 ggplot2_2.2.1    

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.9        nloptr_1.0.4       plyr_1.8.4         iterators_1.0.8   
 [5] tools_3.3.2        digest_0.6.12      lme4_1.1-12        jsonlite_1.3      
 [9] tibble_1.2         gtable_0.2.0       nlme_3.1-128       mgcv_1.8-15       
[13] Matrix_1.2-7.1     foreach_1.4.3      DBI_0.5-1          parallel_3.3.2    
[17] SparseM_1.74       stringr_1.2.0      MatrixModels_0.4-1 htmlwidgets_0.8   
[21] stats4_3.3.2       grid_3.3.2         nnet_7.3-12        R6_2.2.0          
[25] minqa_1.2.4        reshape2_1.4.2     car_2.1-4          magrittr_1.5      
[29] scales_0.4.1       codetools_0.2-15   ModelMetrics_1.1.0 htmltools_0.3.5   
[33] MASS_7.3-45        splines_3.3.2      assertthat_0.1     pbkrtest_0.4-6    
[37] colorspace_1.3-2   quantreg_5.29      stringi_1.1.2      visNetwork_1.0.3  
[41] lazyeval_0.2.0     munsell_0.4.3

**Comment by matt32106** [comment]
things i would check:
- if you run your prediction (same one as on the test set) on the validation set, do you find 86%? there could be an index problem somewhere
- is your split between train/val/test really random? (dataset can be biased)

**Comment by prokope3** [comment]
1) Yes
2) I am using caret's createDataPartition function to do my splitting. If I am not mistaken, that should do randomized splitting.

I tried plot all three errors depending on iteration (I saved model every 5 iterations) and behavior of testing error got even weirder (at least for me) - testing starts on ~50% accuracy and every iteration goes lower and lower. Training and validation does the opposite but in larger steps - validation accuracy ends on 86 % (change 46% from start) and testing stops 6% below starting value.

**Comment by matt32106** [comment]
maybe check the % of outcome per dataset? is it the same in test as val and train?

**Comment by prokope3** [comment]
Classes are balanced. In the original data set there was 1 dominant class (88 %) and other two had ~6 % observation each, but I undersampled the majority one.

**Comment by jeremiedb** [comment]
Have you tried running the model without the BatchNorm layers? I've had some struggles on occasions with BatchNorm at inference time, so maybe this is where the issue comes from.

**Comment by prokope3** [comment]
Thanks for suggestion, I will try it.

**Comment by prokope3** [comment]
Ok, it is not BatchNorm issue. But I realized, where may be the problem. When I take trained model and apply it on portion of original data (without undersampling of majority class), then I get test error close to the validation error (but with poor precision).
But when I undersample testing data set as well, then I get 44 %.
If anyone have a hint, how to proceed, I would appreciate any suggestions. Otherwise this thread can be closed/deleted, because it is probably not mxnet related.
Anyway, thanks for your time and effort.

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!

**Comment by javi-clear-image-ai** [comment]
I have the same issue...
**TRAIN-VALIDATION:**
json_stats: {**"accuracy_cls": 0.977055**, "eta": "0:00:00", "iter": 89999, **"loss": 0.097330**, "loss_bbox": 0.015025, "loss_cls": 0.075462, "loss_rpn_bbox_fpn2": 0.000000, "loss_rpn_bbox_fpn3": 0.000000, "loss_rpn_bbox_fpn4": 0.000153, "loss_rpn_bbox_fpn5": 0.000106, "loss_rpn_bbox_fpn6": 0.000000, "loss_rpn_cls_fpn2": 0.000099, "loss_rpn_cls_fpn3": 0.000064, "loss_rpn_cls_fpn4": 0.000068, "loss_rpn_cls_fpn5": 0.000033, "loss_rpn_cls_fpn6": 0.000000, "lr": 0.000020, "mb_qsize": 64, "mem": 8932, "time": 0.372048}

**TESTING:**
INFO task_evaluation.py: 186: copypaste: AP,AP50,AP75,APs,APm,APl
INFO task_evaluation.py: 187: copypaste: -1.0000,-1.0000,-1.0000,-1.0000,-1.0000,-1.0000

And when I visualize the testing image results, only 4 out of 1000 show up, and some of the classes go wrong ... I am gonna try to change the visualization threshold from 0.9 to 0.4, to see what happens....

Any help would be very much appreciated
